Update README.md

dskoda1 opened the pull request:
Slight difference in syntax I noticed between building warped & warped2.

dskoda1 committed:
Update README.md

Slight difference in syntax I noticed between building warped & warped2.

wilseypa commented:
david,

thank you for the pull request.  we have updated the README as per your
patch and added a bit more to it as well.  i'm still learning to use github
and now know that i could've just pushed your patch into the system.  thank
you for it.

we are making nice progress with warped2 and i believe it's not fully
functional on multi-core boxes and clusters.  we will be incorporating more
advanced pending event features over the next few months.  the
warped2-models archive also has some new models.

regards,
p. wilsey

regards,

## p. wilsey

Philip A. Wilsey                        513 556-4779 (voice)
wilseypa@gmail.com
Experimental Computing Lab
eecs.ceas.uc.edu/~paw/ http://secs.ceas.uc.edu/~paw/
EECS Dept, PO Box 210030      513 556-7326 (fax)
Cincinnati, OH  45221-0030          ---- gpg/pgp keys available ----

On Sat, Apr 11, 2015 at 11:33 AM, David Skoda notifications@github.com
wrote:

> ## Slight difference in syntax I noticed between building warped & warped2.
> 
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/wilseypa/warped2/pull/1
> Commit Summary
> - Update README.md
> 
> File Changes
> - _M_ README.md
>   https://github.com/wilseypa/warped2/pull/1/files#diff-0 (2)
> 
> Patch Links:
> - https://github.com/wilseypa/warped2/pull/1.patch
> - https://github.com/wilseypa/warped2/pull/1.diff
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/wilseypa/warped2/pull/1.

dskoda1 commented:
Hi Phil,

Your welcome for the small update to the readme. I do have a question for
you however; I've been running some tests with warped2/Phold in timewarp
mode on a 48 core machine we have here at Binghamton, and getting mixed
results. Based on the code you currently have, what would you expect to be
the optimal configuration of number of worker threads and schedulers to
minimize execution time? I understand it's not fully functional yet so I'm
curious as to what you think of this.

Thank you,

David Skoda
Binghamton University '16
Computer Science

On Tue, Apr 14, 2015 at 1:11 PM, Philip A. Wilsey notifications@github.com
wrote:

> david,
> 
> thank you for the pull request. we have updated the README as per your
> patch and added a bit more to it as well. i'm still learning to use github
> and now know that i could've just pushed your patch into the system. thank
> you for it.
> 
> we are making nice progress with warped2 and i believe it's not fully
> functional on multi-core boxes and clusters. we will be incorporating more
> advanced pending event features over the next few months. the
> warped2-models archive also has some new models.
> 
> regards,
> p. wilsey
> 
> regards,
> 
> ## p. wilsey
> 
> Philip A. Wilsey 513 556-4779 (voice)
> wilseypa@gmail.com
> Experimental Computing Lab
> eecs.ceas.uc.edu/~paw/ http://secs.ceas.uc.edu/~paw/
> EECS Dept, PO Box 210030 513 556-7326 (fax)
> Cincinnati, OH 45221-0030 ---- gpg/pgp keys available ----
> 
> On Sat, Apr 11, 2015 at 11:33 AM, David Skoda notifications@github.com
> wrote:
> 
> > ## Slight difference in syntax I noticed between building warped & warped2.
> > 
> > You can view, comment on, or merge this pull request online at:
> > 
> > https://github.com/wilseypa/warped2/pull/1
> > Commit Summary
> > - Update README.md
> > 
> > File Changes
> > - _M_ README.md
> >   https://github.com/wilseypa/warped2/pull/1/files#diff-0 (2)
> > 
> > Patch Links:
> > - https://github.com/wilseypa/warped2/pull/1.patch
> > - https://github.com/wilseypa/warped2/pull/1.diff
> > 
> > —
> > Reply to this email directly or view it on GitHub
> > https://github.com/wilseypa/warped2/pull/1.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/wilseypa/warped2/pull/1#issuecomment-92983946.

## 

David Skoda
Binghamton University '16
Computer Science

wilseypa commented:
david,

not sure.  we haven't been using/evaluating the phold model.  sounak, can
you look into it and see if it's operating properly?  we have been focusing
our studies on four models: epidemic (from kalyan's papers), pcs, traffic,
and airport (from the ROSS simulator).  personally i have a rather large
disdain for the phold model and don't let the students work with it much.
that said, i suppose it is important for others for us to have it fully
operational.

now, as to your results on the 48-core box: this can be a number of
factors.  out of the box, our system is setup/optimized/tested for uniform
memory access systems with core counts below (approx) 8 and especially SMP
processors.  the system has nice features for partitioning and
specifically, profile based partitioning for clusters (ultimately one of
our principle targets is clusters of multi-/many-core nodes).

warped2 is setup to follow the critical path of events in the simulation.
basically the LTSF queue holds (as a sorted list) the lowest timestamped
events from all LPs in the system.  the worker threads pull from it.  we
had a couple of hickups with fossil collection that could perturb the
worker threads; we are looking into that.  with 48-cores, the single LTSF
queue incurs significant contention and so you have to move to multiple
LTSF queues.  i would probably suggest 4 LTSF queues with worker threads
from each processor bound to a common LTSF queue (to maximize memory
locality).  of course the problem with this is that there has to be a
partitioning of the LPs among the LTSF queues.  this makes it much more
difficult to track the critical path and will generally lead to higher
rollbacks (on a small SMP system, you should see nearly zero rollbacks).
ANYWAY: there are several possible solutions (none yet implemented in
warped2), namely: (i) load balancing, (ii) profile/application based
partitioning of LPs to LTSF queues, and (iii) LP hopping between the LTSF
queues.  in our studies with warped on the 48-core box, the last does not
work well (although it does pretty well on the smaller i7 SMP boxes that we
have locally).  we assume the problem is memory locality.  the first (load
balancing) is quite difficult to control and i still haven't seen it work
well and i still don't have a decent hypothesis on how to make it do so.
that leaves partitioning.  currently the partitioning code in warped2 is to
partition LPs to the heavyweight MPI processes across the cluster.  i do
not know if/how difficult it would be to do this for threads.  we haven't
yet talked about that step.  in the sort term, you could try to tread the 4
processors on the 48-core box as separate nodes and run the partitioning
code in that way.

of course this is just a guess given what you've  (haven't) told me
:-)....basically i'm guessing.  if you can give me more details on the
configurations and results you're getting, i might be able to say more
(although i'm not guaranteeing that).

honestly even with all these comments, the actual problem you might be
having is with the model itself.  as i said, we simply don't use it.

regards,
p. wilsey

regards,

## p. wilsey

Philip A. Wilsey                        513 556-4779 (voice)
wilseypa@gmail.com
Experimental Computing Lab
eecs.ceas.uc.edu/~paw/ http://secs.ceas.uc.edu/~paw/
EECS Dept, PO Box 210030      513 556-7326 (fax)
Cincinnati, OH  45221-0030          ---- gpg/pgp keys available ----

On Tue, Apr 14, 2015 at 7:08 PM, David Skoda notifications@github.com
wrote:

> Hi Phil,
> 
> Your welcome for the small update to the readme. I do have a question for
> you however; I've been running some tests with warped2/Phold in timewarp
> mode on a 48 core machine we have here at Binghamton, and getting mixed
> results. Based on the code you currently have, what would you expect to be
> the optimal configuration of number of worker threads and schedulers to
> minimize execution time? I understand it's not fully functional yet so I'm
> curious as to what you think of this.
> 
> Thank you,
> 
> David Skoda
> Binghamton University '16
> Computer Science
> 
> On Tue, Apr 14, 2015 at 1:11 PM, Philip A. Wilsey <
> notifications@github.com>
> wrote:
> 
> > david,
> > 
> > thank you for the pull request. we have updated the README as per your
> > patch and added a bit more to it as well. i'm still learning to use
> > github
> > and now know that i could've just pushed your patch into the system.
> > thank
> > you for it.
> > 
> > we are making nice progress with warped2 and i believe it's not fully
> > functional on multi-core boxes and clusters. we will be incorporating
> > more
> > advanced pending event features over the next few months. the
> > warped2-models archive also has some new models.
> > 
> > regards,
> > p. wilsey
> > 
> > regards,
> > 
> > ## p. wilsey
> > 
> > Philip A. Wilsey 513 556-4779 (voice)
> > wilseypa@gmail.com
> > Experimental Computing Lab
> > eecs.ceas.uc.edu/~paw/ http://secs.ceas.uc.edu/~paw/
> > EECS Dept, PO Box 210030 513 556-7326 (fax)
> > Cincinnati, OH 45221-0030 ---- gpg/pgp keys available ----
> > 
> > On Sat, Apr 11, 2015 at 11:33 AM, David Skoda notifications@github.com
> > wrote:
> > 
> > > Slight difference in syntax I noticed between building warped &
> > > 
> > > ## warped2.
> > > 
> > > You can view, comment on, or merge this pull request online at:
> > > 
> > > https://github.com/wilseypa/warped2/pull/1
> > > Commit Summary
> > > - Update README.md
> > > 
> > > File Changes
> > > - _M_ README.md
> > >   https://github.com/wilseypa/warped2/pull/1/files#diff-0 (2)
> > > 
> > > Patch Links:
> > > - https://github.com/wilseypa/warped2/pull/1.patch
> > > - https://github.com/wilseypa/warped2/pull/1.diff
> > > 
> > > —
> > > Reply to this email directly or view it on GitHub
> > > https://github.com/wilseypa/warped2/pull/1.
> > 
> > —
> > Reply to this email directly or view it on GitHub
> > https://github.com/wilseypa/warped2/pull/1#issuecomment-92983946.
> 
> ## 
> 
> David Skoda
> Binghamton University '16
> Computer Science
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/wilseypa/warped2/pull/1#issuecomment-93105763.
